source/manual/alerts/app-healthcheck-not-ok.html.md

```diff
@@ -0,0 +1,112 @@
+---
+owner_slack: "#govuk-2ndline"
+title: App healthcheck not ok
+parent: "/manual.html"
+layout: manual_layout
+section: Icinga alerts
+last_reviewed_on: 2020-09-11
+review_in: 6 months
+---
+
+Many apps on GOV.UK have a healthcheck endpoint.
+
+- This is usually `/healthcheck` [[1](https://github.com/alphagov/govuk-puppet/blob/2693343ebc1aced7a7f94e8aba31fee8b05df8a5/modules/govuk/manifests/apps/email_alert_api.pp#L166)].
+- Some apps just pick a random page [[1](https://github.com/alphagov/govuk-puppet/blob/2693343ebc1aced7a7f94e8aba31fee8b05df8a5/modules/govuk/manifests/apps/collections.pp#L51)].
+
+The alert works by [making a request for the healthcheck endpoint](https://github.com/alphagov/govuk-puppet/blob/fab936cb82be7fad42636fcafca3718a8368ebfe/modules/icinga/files/usr/lib/nagios/plugins/check_json_healthcheck#L155) on the machine where the app is running. If you need to test the healthcheck endpoint manually, you can SSH on to the machine and `curl` it yourself.
+
+```
+# SSH on to a machine running Content Publisher
+gds govuk connect ssh -e integration backend
+
+# Find the port it's running on
+ps -ef | grep content-publisher | grep master
+
+# Do the Icinga check manually
+curl localhost:3221/healthcheck
+```
+
+Apps with a custom healthcheck endpoint often make use of [the generic checks in govuk_app_config](https://github.com/alphagov/govuk_app_config/blob/master/docs/healthchecks.md). Some apps also implement [custom checks](https://github.com/alphagov/content-publisher/blob/2a6e68e5161cde6f8ee4329deee9a242f6d04668/app/controllers/healthcheck_controller.rb#L8), and the alert should link to custom documentation to explain these.
+
+## Connection Refused Error
+
+This means the app is not accepting requests for the healthcheck endpoint, and is probably down.
+
+- Check [the processes are running](check-process-running.html).
+- Try the healthcheck endpoint manually, as above.
+
+## Timeout Error
+
+This means the app is responding to requests, but the request is taking too long (over [20 seconds](https://github.com/alphagov/govuk-puppet/blob/fab936cb82be7fad42636fcafca3718a8368ebfe/modules/icinga/files/usr/lib/nagios/plugins/check_json_healthcheck#L155)).
```

Comment:
> This means the app is responding to requests

I don't think this is necessarily true. In the recent incident where email-alert-api was failing to start the healthcheck was timing out even though the app wasn't (couldn't) respond to requests. I believe the healthcheck requests made it into Nginx but then timed out waiting to communicate with the app.

Maybe just remove that statement ?

Reply:
That's true. I've amended to say "accepting" requests, but not necessarily responding: the unicorn worker starts and opens a connection, but then encounters a fatal error and exits (hence the timeout). 

I've also added an extra "check the logs" step to cover this.